omf/fileio/geoh5.py

```diff
@@ -0,0 +1,510 @@
+from __future__ import annotations
+
+import warnings
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from pathlib import Path
+
+import numpy as np
+from geoh5py.data import Data
+from geoh5py.groups import RootGroup
+from geoh5py.objects import BlockModel, Curve, Grid2D, Points, Surface
+from geoh5py.shared import Entity
+from geoh5py.workspace import Workspace
+
+from omf.base import Project, UidModel
+from omf.data import ScalarArray, ScalarData
+from omf.lineset import LineSetElement, LineSetGeometry
+from omf.pointset import PointSetElement, PointSetGeometry
+from omf.surface import SurfaceElement, SurfaceGeometry, SurfaceGridGeometry
+from omf.volume import VolumeElement, VolumeGridGeometry
+
+
+class OMFtoGeoh5NotImplemented(Exception):
+    """Custom error message for rotation not implemented by geoh5."""
+
+    def __init__(
+        self,
+        name: str,
+    ):
+        super().__init__(OMFtoGeoh5Implementation.message(name))
+
+    @staticmethod
+    def message(info):
+        """Custom error message."""
+        return f"Cannot perform the conversion from OMF to geoh5. {info}"
+
+
+class GeoH5Writer:
+    """
+    OMF to geoh5 class converter
+    """
+
+    def __init__(self, element, file_name: str | Path):
+        self._file = None
+        self.file = file_name
+        self.entity = element
+
+    @property
+    def file(self):
+        """Target file on disk."""
+        return self._file
+
+    @file.setter
+    def file(self, file_name: str | Path):
+        if not isinstance(file_name, (str, Path)):
+            raise ValueError("Input 'file' must be of str or Path.")
+
+        self._file = file_name
+
+    @property
+    def entity(self):
+        return self._entity
+
+    @entity.setter
+    def entity(self, element):
+        if type(element) not in _CONVERSION_MAP:
+            raise ValueError(
+                "Element of type {type(element)} currently not implemented."
+            )
+
+        converter: BaseConversion = _CONVERSION_MAP[type(element)](element, self.file)
+        self._entity = converter.from_omf()
+
+
+class BaseConversion(ABC):
+    """
+    Base conversion between OMF and geoh5 format.
+    """
+
+    arguments: dict = {}
+    geoh5: str | Path | Workspace = None
+    geoh5_type = Entity
+    omf_type: type[UidModel] = UidModel
+    _attribute_map: dict = {
+        "uid": "uid",
+        "description": "description",
+        "name": "name",
+    }
+    _element = None
+    _entity = None
+
+    def __init__(self, obj: UidModel | Entity, geoh5: str | Path | Workspace):
+        if isinstance(obj, self.omf_type):
+            self.element = obj
+        elif isinstance(obj, self.geoh5_type):
```

Comment:
The goal is to later use the converters to go the other way too. So if you supply a geoh5 object it would call the from_geoh5 method. Maybe simpler to split the from and to?